src/Cardano/Wallet/Binary/Helpers.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE RankNTypes #-}
+
+-- | Some extra helpers on top of the CBOR library to help writing decoders.
+
+module Cardano.Wallet.Binary.Helpers
+    ( inspectNextToken
+    , decodeList
+    , decodeListIndef
+    , unsafeDeserialiseFromBytes
+    ) where
+
+import           Prelude
+import qualified Codec.CBOR.Decoding              as CBOR
+import qualified Codec.CBOR.Read                  as CBOR
+import qualified Data.ByteString.Lazy             as BL
+import           Debug.Trace                      (traceShow)
+
+-- | Inspect the next token that has to be decoded and print it to the console
+-- as a trace. Useful for debugging Decoders.
+-- Example:
+--
+--     myDecoder :: CBOR.Decoder s MyType
+--     myDecoder = do
+--         a <- CBOR.decodeWord64
+--         inspectNextToken
+--         [...]
+--
+inspectNextToken :: CBOR.Decoder s ()
+inspectNextToken =
+  CBOR.peekTokenType >>= flip traceShow (return ())
+
+-- | Decode an list of known length. Very similar to @decodeListIndef@.
+--
+-- Example:
+--
+--     myDecoder :: CBOR.Decoder s [MyType]
+--     myDecoder = decodeList decodeOne
+--       where
+--         decodeOne :: CBOR.Decoder s MyType
+--
+decodeList :: forall s a . CBOR.Decoder s a -> CBOR.Decoder s [a]
+decodeList decodeOne = do
+    l <- CBOR.decodeListLenCanonical
+    CBOR.decodeSequenceLenN (flip (:)) [] reverse l decodeOne
+
+-- | Decode an arbitrary long list. CBOR introduce a "break" character to
+-- mark the end of the list, so we simply decode each item until we encounter
+-- a break character.
+--
+-- Example:
+--
+--     myDecoder :: CBOR.Decoder s [MyType]
+--     myDecoder = decodeListIndef decodeOne
+--       where
+--         decodeOne :: CBOR.Decoder s MyType
+--
+decodeListIndef :: forall s a. CBOR.Decoder s a -> CBOR.Decoder s [a]
+decodeListIndef decodeOne = do
+    _ <- CBOR.decodeListLenIndef
+    CBOR.decodeSequenceLenIndef (flip (:)) [] reverse decodeOne
+
+-- | CBOR deserialise without error handling - handy for prototypes or testing.
+unsafeDeserialiseFromBytes :: (forall s. CBOR.Decoder s a) -> BL.ByteString -> a
+unsafeDeserialiseFromBytes decoder bytes =
+    either (\e -> error $ "unsafeDeserialiseFromBytes: " <> show e) snd $
+        CBOR.deserialiseFromBytes decoder bytes
```

Comment:
:thinking:  ... I'd rather _not_ have that one in our main source files; this was handy for testing and quick prototyping but not very type-safe; in practice, we'll definitely want to report on that error differently.

Reply:
Fair enough, done.